Modules/_testcapi/watchers.c

```diff
@@ -277,6 +278,149 @@ unwatch_type(PyObject *self, PyObject *args)
+
+static int
+handle_code_object_event(int which_watcher, PyCodeEvent event, PyCodeObject *co) {
+    if (event == PY_CODE_EVENT_CREATE) {
+        num_code_object_created_events[which_watcher]++;
+    }
+    else if (event == PY_CODE_EVENT_DESTROY) {
+        num_code_object_destroyed_events[which_watcher]++;
+    }
+    else {
+        return -1;
+    }
+    return 0;
+}
+
+static int
+first_code_object_callback(PyCodeEvent event, PyCodeObject *co)
+{
+    return handle_code_object_event(0, event, co);
+}
+
+static int
+second_code_object_callback(PyCodeEvent event, PyCodeObject *co)
+{
+    return handle_code_object_event(1, event, co);
+}
+
+static int
+noop_code_event_handler(PyCodeEvent event, PyCodeObject *co)
+{
+    return 0;
+}
+
+static PyObject *
+register_code_watcher(PyObject *self, PyObject *which_watcher)
+{
+    int watcher_id;
+    assert(PyLong_Check(which_watcher));
+    long which_l = PyLong_AsLong(which_watcher);
+    if (which_l == 0) {
+        watcher_id = PyCode_RegisterWatcher(first_code_object_callback);
+    }
+    else if (which_l == 1) {
+        watcher_id = PyCode_RegisterWatcher(second_code_object_callback);
+    }
+    else {
+        return NULL;
+    }
+    if (watcher_id < 0) {
+        return NULL;
+    }
+    return PyLong_FromLong(watcher_id);
+}
+
+static PyObject *
+clear_code_watcher(PyObject *self, PyObject *watcher_id)
+{
+    assert(PyLong_Check(watcher_id));
+    long watcher_id_l = PyLong_AsLong(watcher_id);
+    if (PyCode_ClearWatcher(watcher_id_l) < 0) {
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
+
+static PyObject *
+get_code_watchers_events_counts()
+{
+    PyObject *res = PyDict_New();
+    if (res == NULL) {
+        Py_RETURN_NONE;
+    }
+    for (int i=0; i < NUM_CODE_WATCHERS; i++) {
+        PyObject *key = PyLong_FromLong(i);
+        PyObject *value = PyDict_New();
+        PyObject *num_created = PyLong_FromLong(num_code_object_created_events[i]);
+        PyObject *num_destroyed = PyLong_FromLong(num_code_object_destroyed_events[i]);
+        if (key == NULL || value == NULL ||
+            num_created == NULL || num_destroyed == NULL) {
+            Py_XDECREF(key);
+            Py_XDECREF(value);
+            Py_XDECREF(num_created);
+            Py_XDECREF(num_destroyed);
+            return NULL;
+        }
+        if ((PyDict_SetItemString(value, "created", num_created) < 0) ||
+            (PyDict_SetItemString(value, "destroyed", num_destroyed) < 0)) {
+            Py_DECREF(key);
+            Py_DECREF(value);
+            Py_DECREF(num_created);
+            Py_DECREF(num_destroyed);
+            return NULL;
+        }
+        Py_DECREF(num_created);
+        Py_DECREF(num_destroyed);
+        if (PyDict_SetItem(res, key, value) < 0) {
+            Py_DECREF(key);
+            Py_DECREF(value);
+            return NULL;
+        }
+        Py_DECREF(key);
+        Py_DECREF(value);
+    }
+    return res;
+}
```

Comment:
nit: it might be somewhat simpler / less boilerplate C-API code to have this method take a watcher id and a boolean for created/destroyed and just return a single count; let a Python wrapper handle calling with the full matrix and creating a dict.